### PR TITLE
Re-write the documentation index

### PR DESCRIPTION
**Pull request by steveklabnik**
The docs team has decided that we're framing resources in three ways:
"learning Rust," "using Rust," "mastering Rust." This is a more useful
split than "beginner/intermediate/advanced." As we add more resources
in the future, we expect "using Rust" to grow. "the bookshelf" as a
concept is great, but isn't really organized along these lines. As such,
this reorganizes the docs along these lines.

r? @rust-lang/docs

### CODE REVIEW

**Comment by TimNN** [comment]
Your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/361725962). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:00:55] configure: rust.quiet-tests     := True
---
[00:40:08] ..........................................................................i.........................
[00:40:14] .................i..................................................................................
---
[00:40:49] .............................................................................................i......
[00:40:56] .................................................................i..................................
---
[00:41:51] .............................................i......................................................
---
[00:45:43] .............................i......................................................................
[00:45:57] ..............................................................i.....................................
[00:46:13] ...............................................i....................................................
[00:46:33] ....................................................................................................
[00:46:54] ....................................................................................................
[00:47:16] ....................................................................................................
[00:47:41] .i...............................................................................................i..
[00:48:08] .................................................................................test [run-pass] run-pass/mir_heavy_promoted.rs has been running for over 60 seconds
[00:48:18] ...................
[00:48:48] ....................................................................................................
[00:49:23] ..............................................................ii....................................
[00:50:09] .........................i....................................................i.ii.....test [run-pass] run-pass/saturating-float-casts.rs has been running for over 60 seconds
[00:50:13] .............
[00:50:53] ......................................................................................iiiiiii.......
---
[00:52:53] ....................................i...............................................................
[00:53:01] ....................................................................................................
[00:53:08] ..................i............................................................ii.iii...............
[00:53:15] ....................................................................................................
[00:53:23] ........i..............................i............................................................
[00:53:30] ....................................................................................................
[00:53:37] .....................i..............................................................................
[00:53:45] ....................................................................................................
[00:53:55] ....................................................................................................
[00:54:05] ....................................................................................................
[00:54:16] ....................................................................................................
[00:54:30] ....................................................................................................
[00:54:38] ..............i.....................................................................................
[00:54:48] .................i..ii..............................................................................
[00:54:58] ....................................................................................................
[00:55:08] ....................................................................................................
[00:55:18] .....................................................................................i..............
[00:55:28] ...............................i....................................................................
---
[00:56:05] ...........................i........................................................................
[00:56:07] ....................................................................i...............................
[00:56:08] ................i.......................................................
---
[00:56:22] ...........i........................
---
[00:56:52] i....i.ii....i.............ii........iii......i..i...i...ii..i..i..ii.....
---
[00:56:55] i.......i......................i......
---
[00:57:31] iiii.......i..i........i..i.i.............i..........iiii...........i...i..........ii.i.i.......ii..
[00:57:32] ....ii...
---
[01:06:38] ...i................................................................................................
---
[01:08:27] ......................................i.............................................................
[01:08:46] ....................................................................................................
[01:09:06] .............................................i......................................................
---
[01:10:33] ..........................................................ii........................................
---
[01:11:35] ..............................................................i.....................................
---
[01:16:13] ii..................................................................................................
[01:16:31] ....................................................................................................
[01:16:47] ...................iii......i......i...i......i.....................................................
[01:16:56] ....................................................................................................
[01:17:11] ........................................iiii........ii..............................................
[01:17:22] ....................................................................................................
[01:17:38] .......................................................................................i............
---
[01:24:02] index.html:80: broken link - err-index.html
[01:24:18] thread 'main' panicked at 'found some broken links', tools/linkchecker/main.rs:49:9
---
[01:24:18] Makefile:58: recipe for target 'check' failed
[01:24:18] make: *** [check] Error 1
---
$ cat obj/tmp/sccache.log
```

</details><p></p>

I'm a bot! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**.

File: src/doc/index.md [hunk 1]
```diff
@@ -6,55 +6,72 @@ nav {
...
-[refchecklist]: https://github.com/rust-lang-nursery/reference/issues/9
-[err]: error-index.html
-[book]: book/index.html
-[rbe]: rust-by-example/index.html
-[nomicon]: nomicon/index.html
-[unstable-book]: unstable-book/index.html
-[rustdoc-book]: rustdoc/index.html
-[cargo-book]: cargo/index.html
+diagnostics from the compiler on those errors. You can also [read them
+here](err-index.html), if you prefer to read them that way.
```
Comment (QuietMisdreavus) [inline src/doc/index.md]: Link should point to `error-index.html`

**Comment by QuietMisdreavus** [comment]
Overall, this gets a :+1: from me, though it's kind of an overhaul and likely to generate lots of Opinions. I'd like to see what the rest of @rust-lang/docs thinks.

**Comment by steveklabnik** [comment]
Let's do it this way: 

@rfcbot fcp merge

**Comment by steveklabnik** [comment]
Let's do it this way: 

@rfcbot fcp merge

**Comment by steveklabnik** [comment]
Checked of @QuietMisdreavus as per 👍 above

**Review by frewsxcv — APPROVED** [review]
looks great, love these new categories ✨

**Comment by GuillaumeGomez** [comment]
:+1:

**Comment by steveklabnik** [comment]
@bors: r=QuietMisdreavus 

thanks everyone!

**Pull request merged**